cmd/notation/inspect.go

```diff
@@ -0,0 +1,294 @@
+}
+
+func logSkippedSignature(sigDesc ocispec.Descriptor, err error) {
+	fmt.Fprintf(os.Stderr, "Warning: Skipping signature %s because of error: %v\n", sigDesc.Digest.String(), err)
+}
+
+func getSignedAttributes(outputFormat string, envContent *signature.EnvelopeContent) map[string]string {
+	signedAttributes := map[string]string{
+		"signingScheme": string(envContent.SignerInfo.SignedAttributes.SigningScheme),
+		"signingTime":   formatTimestamp(outputFormat, envContent.SignerInfo.SignedAttributes.SigningTime),
+		"expiry":        formatTimestamp(outputFormat, envContent.SignerInfo.SignedAttributes.Expiry),
+	}
+
+	for _, attribute := range envContent.SignerInfo.SignedAttributes.ExtendedAttributes {
+		signedAttributes[fmt.Sprint(attribute.Key)] = fmt.Sprint(attribute.Value)
+	}
+
+	return signedAttributes
+}
+
+func getUnsignedAttributes(envContent *signature.EnvelopeContent) map[string]string {
+	unsignedAttributes := map[string]string{}
+
+	if envContent.SignerInfo.UnsignedAttributes.TimestampSignature != nil {
+		unsignedAttributes["timestampSignature"] = b64.StdEncoding.EncodeToString(envContent.SignerInfo.UnsignedAttributes.TimestampSignature)
+	}
+
+	if envContent.SignerInfo.UnsignedAttributes.SigningAgent != "" {
+		unsignedAttributes["signingAgent"] = envContent.SignerInfo.UnsignedAttributes.SigningAgent
+	}
+
+	return unsignedAttributes
+}
+
+func formatTimestamp(outputFormat string, t time.Time) string {
+	switch outputFormat {
+	case cmd.OutputJSON:
+		return t.Format(time.RFC3339)
+	default:
+		return t.Format(time.ANSIC)
+	}
+}
+
+func getCertificates(outputFormat string, envContent *signature.EnvelopeContent) []certificateOutput {
+	certificates := []certificateOutput{}
+
+	for _, cert := range envContent.SignerInfo.CertificateChain {
+		h := sha1.Sum(cert.Raw)
+		fingerprint := strings.ToLower(hex.EncodeToString(h[:]))
+
+		certificate := certificateOutput{
+			SHA1Fingerprint: fingerprint,
+			IssuedTo:        cert.Subject.String(),
+			IssuedBy:        cert.Issuer.String(),
+			Expiry:          formatTimestamp(outputFormat, cert.NotAfter),
+		}
+
+		certificates = append(certificates, certificate)
+	}
+
+	return certificates
+}
+
+func printOutput(outputFormat string, ref string, output inspectOutput) error {
+	if outputFormat == cmd.OutputJSON {
+		return ioutil.PrintObjectAsJSON(output)
+	}
+
+	fmt.Println("Inspecting all signatures for signed artifact")
+	root := tree.New(ref)
+	root.Add(NotarySignatureMediaType)
+
+	for _, signature := range output.Signatures {
+		root.AddPath(NotarySignatureMediaType, signature.Digest, fmt.Sprintf("media type: %s", signature.MediaType))
+		root.AddPath(NotarySignatureMediaType, signature.Digest, fmt.Sprintf("signature algorithm: %s", signature.SignatureAlgorithm))
+
+		for k, v := range signature.SignedAttributes {
+			root.AddPath(NotarySignatureMediaType, signature.Digest, "signed attributes", fmt.Sprintf("%s: %s", k, v))
+		}
+
+		for k, v := range signature.UserDefinedAttributes {
```

Comment:
Looks like when the signature does not have any user defined attribute, the phrase `user defined attributes` would still be printed out. Shall we just skip the phrase in this case?
```
  │   │   ├── expiry: 0001-01-01 00:00:00 +0000 UTC
    │   │   └── signingScheme: notary.x509
    │   ├── user defined attributes
    │   ├── unsigned attributes
    │   │   └── signingAgent: Notation/1.0.0
```

Reply:
We can keep as it is or just show `(empty)`.

Reply:
will change to `(empty)`, i personally prefer having some indicator that the field is present without any value